### ISSUE TITLE
build: create independent conda-forge packages to backends separately of the main one

### ISSUE THREAD

**Issue by datapythonista** [ibis-project/ibis#2448]
We've been having problems with the dependencies of one backend conflicting with the dependencies of other backends. This has created conflicts (see #2426) and also makes the resolving of the dependencies very slow (see #2324).

One of the solutions is to move some of the backends to separate repos, and package them independently of the Ibis core (see #2356). Regardless of whether we move backends to separate repos, having independent conda packages would be convenient to avoid these problems.

We can probably start creating the packages for the trickier ones (omnisci, bigquery), and leave the simpler for later, since there are a significant amount of backends, and it'll be a decent amount of work.

**Comment by jreback** [comment]
see my comments elsewhere


this is a massive increase in operational complexity

spitting a single package and seeing how it goes is probably ok (omnisc) - and just to be clear this is just a recipe split - or are you proposing a separate repo?

we don't want anything manual

**Comment by datapythonista** [comment]
I thought you proposed this one (having multiple conda-forge packages). What would you do then?

**Comment by jreback** [comment]
yeah multiple conda forge packages but a single repo (as we have now)

we already split the testing so the conda packages should mirror this

docs should also only have the ibis core packages (otherwise cannot solve easily)

but i don't really see this as a problem - you don't need the specific backends for the docs

**Comment by datapythonista** [comment]
Ok, didn't know a single conda-forge repo could generate several packages, but much better if that's the case.

I'll have a look at the docs. Would be great if we have a way to build the docs without having to import all the dependencies. If we find it, maybe it can also work for adding docs from external repos. In any case, not really important I think.

**Comment by matthewmturner** [comment]
im new to conda packaging (from a dev perspective) but happy to lend a hand here.  Is there anything in particular that you could use help on with this?

**Comment by datapythonista** [comment]
> im new to conda packaging (from a dev perspective) but happy to lend a hand here. Is there anything in particular that you could use help on with this?

Thanks!

I need to do more research, but I think it's mainly updating `ci/recipe/meta.yaml` to only package the core in the main package, and add subpackages for each backend. We first need to update the recipe with the one in https://github.com/conda-forge/ibis-framework-feedstock/ (issue #2523). If you want to take care of that, it should be really simply (just copy the file and leave the comment in the header).

And once that is that done, if you want to have a look at starting by moving one backend as a subpackage in the recipe, that would be great.

**Comment by matthewmturner** [comment]
Is this the approach being taken for the subpackages?

https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html#implicit-metapackages

**Comment by datapythonista** [comment]
> Is this the approach being taken for the subpackages?
> 
> https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html#implicit-metapackages

That's what I found. @jreback can you confirm this is what you had in mind when you said we can release multiple packages with a single conda-forge repo?

**Comment by jreback** [comment]
no we need to do something different. here is one way

we have a package: ibis-core
this just packages minimal dependencies (say pandas)

the we have packages that are like this

ibis-omnisci (depends on omnisci stuff + ibis-core)
ibis-sql
ibis-spark
ibis-biguery 

each of these has a yaml that is similar to the ci requirements that we have

now how to divide this up
it's a trade off between ease (1 package is easy but is hard to solve) and flexibility (many packages)

so i would start with omnisci and maybe spark the rest might be ok together

**Comment by datapythonista** [comment]
Do you know of any project that has multiple recipe files for reference? Haven't seen in the docs that option.

And since omnisci is being moved to another repo soon, probably better to start with spark.

**Comment by jreback** [comment]
i wouldn't actually move to another repo that complicates things - just a recipe

let's start really simple

**Comment by datapythonista** [comment]
I think there is still some confusion, based on the discussions here, and in #2534.

I think for omnisci, since we are moving the backend to a separate repo in #2356, it's clear that it will have it's own feedstock repo and pypi project, as mssql and arrow do. So, I wouldn't make changes to it, other that removing everything from it in this repo. Also, it's not a great backend to make tests, because `pymapd` (its main dependency) causes all sort of problems with conda, that's one of the reasons we are moving it elsewhere.

For the rest, I assume we will be releasing the base backends with the core ibis, and there is nothing to do about them. This leaves us with the next 13 backends:
- bigquery
- clickhouse
- csv
- dask
- hdf5
- impala
- mysql
- pandas
- parquet
- postgres
- pyspark
- spark
- sqlite

For now I'd keep at least `pandas`, `csv`, `hdf5` and `parquet` in the core ibis package.

If I understand @jreback correctly, what we want to do with the rest or some of them, is to keep their code in the `github.com/ibis-project/ibis` repo, but split its recipe from `github.com/conda-forge/ibis-framework-feedstock/`, into new feedstocks, for example `github.com/conda-forge/ibis-bigquery-feedstock/`. Am I right?

Then, we want to start by a single backend, and see how it goes. I'd start by bigquery (open to another one too).

And we need to decide few things:
- Do we want to also release in pypi separately? Creating an `ibis-bigquery` project, instead of releasing as `ibis[bigquery]`?
- Do we want to test the conda build for bigquery (and the rest of the backends later) in the CI here? Or just testing the core is enough? If we do want to test them, is a single build for building all the conda packages enough, or do we want a build in this repo's CI for each backend?
- What is missing in #2534 is that we are creating a new package with only the backend dependencies, but we're still packing the whole ibis again. I think what it's needed to release a backend separately is that every backend has it's own `setup.py`. Or do you have something else in mind @jreback?

**Comment by jreback** [comment]
let's start really really simple. move omnisci to a separate repo with a separate package & pypyi package that depends on ibis itself.

I think this will solve 99% of the problems. IF and only if we have other issues should we consider doing this to other pieces that are causing issues.

we DO NOT want to create an operational nightmare here. let's not split anything else right now.

In ibis itself we need to remove pymapd from the docs (again what is causing the issues).

I think we should *still* test omnisci on the main ibis CI itself until the split is done (then it simply could be done in the new repo). though even after not averse to still having CI for omnisci in the main repo (it enforces consistency) and if its relatively cheap then no big deal.

so to summarize at the end of this we will have 2 packages (conda & pypi).
- ibis
- ibis-omnisci (depends on ibis)

**Comment by jreback** [comment]
we *could* try *just* splitting out into a separate recipe the omni (and removing it from the docs). this would also work IMHO (and is simpler), rather than setting up *another* repo.

**Comment by matthewmturner** [comment]
@jreback based on your last comment (and if this is the desired approach) of just creating the separate omnisci recipe can we leverage my existing PR #2534 ?

**Comment by cpcloud** [comment]
cc @gforsyth

**Comment by gforsyth** [comment]
Good times.
There's an inherent operational nightmare here because `conda` has no sense of `extras_require` or similar.

I would suggest that we leave the `pypi` package alone, in as as much as it's already possible to select specific backends to install.

For `conda-forge`, we don't want to break everything, but I think we can not break anything but provide extra packages that only target specific backends (or subsets of backends).

I'd propose that `ibis-framework-core` and `ibis-framework` remain as they are, where `core` is the bare bones, and `ibis-framework` is nearly everything.

Then, for those who only want a specific backend (let's say `spark`), we can create (in the existing ibis-framework-feedstock repo) an `ibis-framework-spark` package that only depends on `ibis-framework-core`, `pyarrow`, and `pyspark` and then repeat for the various backends.

All of these can be built in the same conda-forge recipe and no one has to change their install process unless they _want_ to to take advantage of a lighter dependency load.

**Comment by cpcloud** [comment]
100% on board with that approach :100:

**Comment by gforsyth** [comment]
Opened a PR to add a bunch of backend-specific packages -- once that's in I can update the installation documentation to reflect the conda options for single-backend installs

**Comment by gforsyth** [comment]
Ok, there are now packages for:
```
ibis-clickhouse
ibis-dask
ibis-hdf5
ibis-impala
ibis-parquet
ibis-postgres
ibis-pyspark
ibis-sqlite
```
all handled through the `ibis-framework-feedstock` repo.

There are additionally
```
ibis-omniscidb
ibis-vega-transform
ibis-mssql
ibis-bigquery
```

that are handled separately (the `mssql` one is quite old and we might want to adopt it).

Once @cpcloud 's docs PR is merged in, we can add the new install commands.

**Comment by cpcloud** [comment]
Docs are in. Closing this out, thanks @gforsyth!
